README.md
Dissolve
========

Parse and consume binary streams with a neat DSL.

Overview
--------

Dissolve allows you to parse packed binary data into numbers, buffers, strings
and more*! With a simple syntax inspired by [node-binary](https://github.com/substack/node-binary)
and a solid, minimal implementation, you can be up and running in no time.

(* implementing "more" is left as an exercise to the reader)

If you want to produce binary data, might I suggest [concentrate](https://github.com/deoxxa/concentrate)?

Features
--------

* Accurate handling of [u]int{8,16,32} numbers in both signed and unsigned
  variants using fast, built-in [Buffer](http://nodejs.org/docs/latest/api/buffer.html)
  methods
* Fast approximation of [u]int64 numbers in signed and unsigned variants
* Extendable base class for building your own parsers and implementing
  custom types
* Tiny (~250 LoC) implementation, allowing for easy debugging

Installation
------------

Available via [npm](http://npmjs.org/):

> $ npm install dissolve

Or via git:

> $ git clone git://github.com/deoxxa/dissolve.git node_modules/dissolve

Usage
-----

Also see [example.js](https://github.com/deoxxa/dissolve/blob/master/example.js),
[example-complex.js](https://github.com/deoxxa/dissolve/blob/master/example-complex.js)
and [example-loop.js](https://github.com/deoxxa/dissolve/blob/master/example-loop.js).

```javascript
#!/usr/bin/env node

var Dissolve = require("./index");

var parser = Dissolve().loop(function(end) {
  this.uint8("id").tap(function() {
    if (this.vars.id === 0x01) {
      this.uint16be("a").uint16be("b");
    } else if (this.vars.id === 0x02) {
      this.uint32be("x").uint32be("y");
    }
  }).tap(function() {
    this.push(this.vars);
    this.vars = {};
  });
});

parser.on("readable", function() {
  var e;
  while (e = parser.read()) {
    console.log(e);
  }
});

parser.write(new Buffer([0x01, 0x00, 0x02, 0x00, 0x03])); // {id: 1, a: 2, b: 3}
parser.write(new Buffer([0x02, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0x05])); // {id: 2, x: 4, y: 5}
parser.write(new Buffer([0x01]));
parser.write(new Buffer([0x00, 0x02, 0x00]));
parser.write(new Buffer([0x03])); // {id: 1, a: 2, b: 3}
```

Methods
-------

All parser methods are chainable and return the same parser instance they were
called on.

Tap
---

`tap(name, callback)`

This method allows you to "tap" into the parser at an arbitrary point. The
callback will be called bound to the parser instance, so you can use parser
methods on `this`. Any additional parser steps you introduce inside the callback
will be executed before any existing steps that are already scheduled to run
after the `tap` call.

If you provide a `name` parameter, all actions performed in the callback will be
applied to a child object that will be put into a new property named after
`name`. Note that in the callback, even if you provide a `name` parameter, you
can still pretend you were in the outer "scope" because of some prototype
trickery done with the `vars` object under the hood. You don't need to worry
about that too much, the examples should make it a bit clearer.

Loop
----

`loop(name, callback)`

This method is like `tap` except that the callback is called over and over until
signalled to stop. You do this by calling the `end` function that's provided as
the first argument to your callback. When you call the `end` function, you can
provide an optional truthy/non-truthy flag to tell Dissolve to ignore the result
of the iteration of the loop where `end` was called. This is useful if you are
reading until a null entry or similar.

If you provide a `name` parameter, a new array will be placed into a property
named for that parameter, and after each iteration of the loop, any new values
will be appended to the array as an object. As with the `name` stuff on `tap`,
the examples will make that explanation a lot clearer.

The same semantics for job ordering and "scoping" apply as for `tap`.

Basic Parsing Methods
---------------------

For each basic parsing method, the `name` value is the key under which the value
will be attached to `this.vars`.

Buffer/String Methods
---------------------

For these methods, the `length` parameter tells the parser how many bytes to
pull out. If it's a string, it will be assumed that it is the name of a
previously-set `this.vars` entry. If it's a number, it will be used as-is.

* `buffer(name, length)` - binary slice
* `string(name, length)` - utf8 string slice
* `skip(length)` - skip `length` bytes

Numeric Methods
---------------

* `int8(name)` - signed 8 bit integer
* `sint8(name)` - signed 8 bit integer
* `uint8(name)` - unsigned 8 bit integer
* `int16(name)` - signed, little endian 16 bit integer
* `int16le(name)` - signed, little endian 16 bit integer
* `int16be(name)` - signed, big endian 16 bit integer
* `sint16(name)` - signed, little endian 16 bit integer
* `sint16le(name)` - signed, little endian 16 bit integer
* `sint16be(name)` - signed, big endian 16 bit integer
* `uint16(name)` - unsigned, little endian 16 bit integer
* `uint16le(name)` - unsigned, little endian 16 bit integer
* `uint16be(name)` - unsigned, big endian 16 bit integer
* `int32(name)` - signed, little endian 32 bit integer
* `int32le(name)` - signed, little endian 32 bit integer
* `int32be(name)` - signed, big endian 32 bit integer
* `sint32(name)` - signed, little endian 32 bit integer
* `sint32le(name)` - signed, little endian 32 bit integer
* `sint32be(name)` - signed, big endian 32 bit integer
* `uint32(name)` - unsigned, little endian 32 bit integer
* `uint32le(name)` - unsigned, little endian 32 bit integer
* `uint32be(name)` - unsigned, big endian 32 bit integer
* `int64(name)` - signed, little endian 64 bit integer
* `int64le(name)` - signed, little endian 64 bit integer
* `int64be(name)` - signed, big endian 64 bit integer
* `sint64(name)` - signed, little endian 64 bit integer
* `sint64le(name)` - signed, little endian 64 bit integer
* `sint64be(name)` - signed, big endian 64 bit integer
* `uint64(name)` - unsigned, little endian 64 bit integer
* `uint64le(name)` - unsigned, little endian 64 bit integer
* `uint64be(name)` - unsigned, big endian 64 bit integer
* `floatbe(data)` - big endian 32 bit float
* `floatle(data)` - little endian 32 bit float
* `doublebe(data)` - big endian 64 bit double
* `doublele(data)` - little endian 64 bit double

License
-------

3-clause BSD. A copy is included with the source.

Contact
-------

* GitHub ([deoxxa](http://github.com/deoxxa))
* Twitter ([@deoxxa](http://twitter.com/deoxxa))
* ADN ([@deoxxa](https://alpha.app.net/deoxxa))
* Email ([deoxxa@fknsrs.biz](mailto:deoxxa@fknsrs.biz))


CHANGELOG.md

*   Add a new `skip` operation.

    This new operation allows to easily `skip` a specific number of bytes. It
    has the same semantics as `buffer` and `string`.

    *Nikolay Perevozchikov*

*   Refactor `dissolve` internals to improve performance.

    By slicing up the `Dissolve#_transform` method into smaller methods,
    we can allow V8 to optimize these smaller methods individually, while
    making each method less prone to deoptimizations.

    Reduce GC pressure by reducing the number of intermediate arrays that get
    created in `tap` and `loop` operations

    *Arthur Schreiber*

*   Allow `objectMode` to be overridden.

    By default, the `dissolve` stream is in `objectMode`, but disabling this
    mode allows piping a `dissolve` stream into another stream that only
    accepts `Buffer` objects.

    *Lee Treveil*


test/tap.js
var assert = require("chai").assert;
var Dissolve = require("../index");

describe("tap", function() {
  it("should emit data 2 times then end", function(done) {
    var reader = Dissolve().uint8("x").tap(function() {
      this.push(this.vars.x);

      this.uint8("y").tap(function() {
        this.push(this.vars.y);

        this.tap(function() {
          this.push(null);
        });
      });
    });

    var counter = 0;

    reader.on("readable", function() {
      var data;
      while (data = reader.read()) {
        counter++;
      }
    });

    reader.on("end", function() {
      if (counter !== 2) {
        return done(Error("invalid counter value"));
      } else {
        return done();
      }
    });

    reader.write(Buffer([0x01, 0x01]));
  });

  it("should populate child objects correctly", function(done) {
    var reader = Dissolve().tap("a", function() {
      this.uint8("x").tap("b", function() {
        this.uint8("y").tap(function() {
          this.push(this.vars);
        });
      });
    });

    reader.once("readable", function() {
      var e = reader.read();

      assert.isObject(e);
      assert.isObject(e.a);
      assert.isObject(e.a.b);
      assert.strictEqual(e.a.x, 1);
      assert.strictEqual(e.a.b.y, 1);

      assert.isNull(reader.read());

      done();
    });

    reader.write(Buffer([0x01, 0x01]));
  });
});


test/loop.js
var Dissolve = require("../index");

describe("loop", function() {
  it("should emit data 3 times then end", function(done) {
    var reader = Dissolve().loop(function(end) {
      this.uint8("x").tap(function() {
        if (this.vars.x === 0) {
          return end(true);
        } else {
          this.push(this.vars);
        }
      });
    });

    reader.write(Buffer([0x01, 0x01, 0x01, 0x00, 0x02]));

    var counter = 0;

    reader.on("readable", function() {
      var e;
      while (e = reader.read()) {
        counter++;
      }
    });

    reader.on("end", function() {
      if (counter !== 3) {
        return done(Error("invalid counter value"));
      } else {
        return done();
      }
    });
  });

  it("should populate an array correctly", function(done) {
    var reader = Dissolve().loop("things", function(end) {
      this.uint8("x").tap(function() {
        if (this.vars.x === 0) {
          return end(true);
        }
      });
    }).tap(function() {
      this.push(this.vars);
    });

    reader.write(Buffer([0x01, 0x01, 0x01, 0x00]));

    reader.on("readable", function() {
      var e;
      while (e = reader.read()) {
        if (typeof e !== "object" || e === null) {
          return done(Error("invalid payload for data event"));
        }

        if (!Array.isArray(e.things)) {
          return done(Error("array property not set or not correct type"));
        }

        if (e.things.length !== 3) {
          return done(Error("invalid number of entries"));
        }

        return done();
      }
    });
  });

  it("should work with nested, cancelled loop operations", function(done) {
    var reader = Dissolve().tap(function() {
      var data_i = 0;

      this.uint8("data_count").loop("data", function(end) {
        if (data_i++ === this.vars.data_count) {
          return end(true);
        }

        var elements_i = 0;

        this.uint8("element_count").loop("elements", function(end) {
          if (elements_i++ === this.vars.element_count) {
            return end(true);
          }

          this.uint8("element");
        });
      }).tap(function() {
        this.push(this.vars);
        this.vars = Object.create(null);
      });
    });

    reader.write(new Buffer([
      0x02,
        0x02,
          0x01, 0x02,
        0x02,
          0x03, 0x04,
    ]));

    reader.on("readable", function() {
      var e;
      while (e = reader.read()) {
        if (typeof e !== "object" || e === null) {
          return done(Error("invalid payload for data event"));
        }

        if (!Array.isArray(e.data)) {
          return done(Error("array property not set or not correct type"));
        }

        if (e.data.length !== 2) {
          return done(Error("invalid length for data"));
        }

        if (!Array.isArray(e.data[0].elements)) {
          return done(Error("data[0].elements not set or not correct type"));
        }

        if (e.data[0].elements.length !== 2) {
          return done(Error("invalid length for data[0].elements"));
        }

        if (e.data[0].elements[0].element !== 1) {
          return done(Error("invalid value for data[0].elements[0].element"));
        }

        if (e.data[0].elements[1].element !== 2) {
          return done(Error("invalid value for data[0].elements[1].element"));
        }

        if (!Array.isArray(e.data[1].elements)) {
          return done(Error("data[1].elements not set or not correct type"));
        }

        if (e.data[1].elements.length !== 2) {
          return done(Error("invalid length for data[1].elements"));
        }

        if (e.data[1].elements[0].element !== 3) {
          return done(Error("invalid value for data[1].elements[0].element"));
        }

        if (e.data[1].elements[1].element !== 4) {
          return done(Error("invalid value for data[1].elements[1].element"));
        }

        return done();
      }
    });
  });
});


example.js
#!/usr/bin/env node

var Dissolve = require("./index");

var parser = Dissolve().loop(function(end) {
  this.uint8("id").tap("payload", function() {
    this.tap("asdf", function() {
      if (this.vars.id === 0x01) {
        this.uint16be("a").uint16be("b");
      } else if (this.vars.id === 0x02) {
        this.uint32be("x").uint32be("y");
      } else if (this.vars.id === 0x03) {
        this.floatbe("l").doublebe("m");
      }
    });
  }).tap(function() {
    this.emit("data", this.vars);
    this.vars = Object.create(null);
  });
});

parser.on("data", function(e) {
  console.log(e);
});

parser.write(new Buffer([0x01, 0x00, 0x02, 0x00, 0x03])); // {id: 1, a: 2, b: 3}
parser.write(new Buffer([0x02, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0x05])); // {id: 2, x: 4, y: 5}
parser.write(new Buffer([0x01]));
parser.write(new Buffer([0x00, 0x02, 0x00]));
parser.write(new Buffer([0x03])); // {id: 1, a: 2, b: 3}
parser.write(new Buffer([0x03, 0x40, 0x06, 0x66, 0x66, 0x40, 0x00, 0xcc, 0xcc, 0xcc, 0xcc, 0xcc, 0xcd])); // {id: 3, l: ~2.1, m: ~2.1}


complex-example.js
#!/usr/bin/env node

var Dissolve = require("./index"),
    util = require("util");

function Parser() {
  Dissolve.call(this);

  this.loop(function(end) {
    this.uint8("pid").tap(function() {
      switch (this.vars.pid) {
        case 0x00: this.packet_0x00(); break;
        case 0x01: this.packet_0x01(); break;
        case 0x02: this.packet_0x02(); break;
        case 0x03: this.packet_0x03(); break;
        case 0x04: this.packet_0x04(); break;
        case 0xfe: break;
      }
    }).tap(function() {
      this.emit("data", this.vars);
      this.vars = {};
    });
  });
}
util.inherits(Parser, Dissolve);

Parser.prototype.mcstring16 = function string16(name) {
  var len = [name, "len"].join("_");

  return this.uint16be(len).tap(function() {
    this.buffer(name, this.vars[len] * 2).tap(function() {
      delete this.vars[len];

      for (var i=0;i<this.vars[name].length/2;++i) {
        var t = this.vars[name][i*2];
        this.vars[name][i*2] = this.vars[name][i*2+1];
        this.vars[name][i*2+1] = t;
      }

      this.vars[name] = this.vars[name].toString("ucs2");
    });
  });
};

Parser.prototype.packet_0x00 = function packet_0x00() {
  return this.uint32be("token");
};

Parser.prototype.packet_0x01 = function packet_0x01() {
  return this.uint32be("eid").mcstring16("level_type").uint8("game_mode").uint8("dimension").uint8("difficulty").uint8("junk").uint8("max_players");
};

Parser.prototype.packet_0x02 = function packet_0x02() {
  return this.uint8("protocol_version").mcstring16("username").mcstring16("server_host").uint32be("server_port");
};

Parser.prototype.packet_0x03 = function packet_0x03() {
  return this.mcstring16("message");
};

Parser.prototype.packet_0x04 = function packet_0x04() {
  return this.uint64be("time");
};

var parser = new Parser();

parser.on("data", function(e) {
  console.log(e);
});

parser.write(new Buffer([0x00, 0x00, 0x00, 0x00, 0x01]));
parser.write(new Buffer([0x01, 0x00, 0x00, 0x00, 0x01, 0x00, 0x02, 0x00, 0x61, 0x00, 0x62, 0x00, 0x00, 0x00, 0x00, 0x00]));
parser.write(new Buffer([0x02, 0x01, 0x00, 0x02, 0x00, 0x61, 0x00, 0x62, 0x00, 0x02, 0x00, 0x63, 0x00, 0x64, 0x00, 0x00, 0x00, 0x05]));
parser.write(new Buffer([0x03, 0x00, 0x02, 0x00, 0x65, 0x00, 0x66]));
parser.write(new Buffer([0x04, 0x00, 0x00, 0x00, 0x00, 0x01, 0x01, 0x01, 0x01]));


example-complex.js
#!/usr/bin/env node

var Dissolve = require("./index"),
    util = require("util");

function Parser() {
  Dissolve.call(this);

  this.loop(function(end) {
    this.uint8("pid").tap(function() {
      switch (this.vars.pid) {
        case 0x00: this.uint32be("token"); break;
        case 0x01: this.uint32be("eid").mcstring16("level_type").uint8("game_mode").uint8("dimension").uint8("difficulty").uint8("junk").uint8("max_players"); break;
        case 0x02: this.uint8("protocol_version").mcstring16("username").mcstring16("server_host").uint32be("server_port"); break;
        case 0x03: this.mcstring16("message"); break;
        case 0x04: this.uint64be("time"); break;
        case 0xfe: break;
      }
    }).tap(function() {
      this.emit("data", this.vars);
      this.vars = {};
    });
  });
}
util.inherits(Parser, Dissolve);

Parser.prototype.mcstring16 = function string16(name) {
  var len = [name, "len"].join("_");

  return this.uint16be(len).tap(function() {
    this.buffer(name, this.vars[len] * 2).tap(function() {
      delete this.vars[len];

      for (var i=0;i<this.vars[name].length/2;++i) {
        var t = this.vars[name][i*2];
        this.vars[name][i*2] = this.vars[name][i*2+1];
        this.vars[name][i*2+1] = t;
      }

      this.vars[name] = this.vars[name].toString("ucs2");
    });
  });
};

var parser = new Parser();

parser.on("data", function(e) {
  console.log(e);
});

parser.write(new Buffer([0x00, 0x00, 0x00, 0x00, 0x01]));
parser.write(new Buffer([0x01, 0x00, 0x00, 0x00, 0x01, 0x00, 0x02, 0x00, 0x61, 0x00, 0x62, 0x00, 0x00, 0x00, 0x00, 0x00]));
parser.write(new Buffer([0x02, 0x01, 0x00, 0x02, 0x00, 0x61, 0x00, 0x62, 0x00, 0x02, 0x00, 0x63, 0x00, 0x64, 0x00, 0x00, 0x00, 0x05]));
parser.write(new Buffer([0x03, 0x00, 0x02, 0x00, 0x65, 0x00, 0x66]));
parser.write(new Buffer([0x04, 0x00, 0x00, 0x00, 0x00, 0x01, 0x01, 0x01, 0x01]));


index.js
var BufferList = require("bl"),
    stream = require("readable-stream");

var Dissolve = module.exports = function Dissolve(options) {
  if (!(this instanceof Dissolve)) { return new Dissolve(options); }

  if (!options) {
    options = {};
  }

  if (!options.hasOwnProperty('objectMode')) {
    options.objectMode = true;
  }

  stream.Transform.call(this, options);

  this.jobs = [];
  this.vars = Object.create(null);
  this.vars_list = [];

  this._buffer = new BufferList();
};
Dissolve.prototype = Object.create(stream.Transform.prototype, {constructor: {value: Dissolve}});

Dissolve.prototype._job_down = function _job_down(job) {
  var tmp = this.vars;
  this.vars_list.push(tmp);
  this.vars = tmp[job.into] = Object.create(tmp);
};

Dissolve.prototype._job_up = function _job_up() {
  this.vars.__proto__ = null;
  this.vars = this.vars_list.pop();
};

Dissolve.prototype._exec_down = function _exec_down(job) {
  this.jobs.shift();
  this._job_down(job);
};

Dissolve.prototype._exec_up = function _exec_up(job) {
  this.jobs.shift();
  this._job_up();
};

Dissolve.prototype._exec_tap = function _exec_tap(job) {
  this.jobs.shift();

  var jobs = this.jobs;

  this.jobs = [];

  if (job.name) {
    this.jobs.push({type: "down", into: job.name});
    this.jobs.push({type: "tap", args: job.args, fn: job.fn});
    this.jobs.push({type: "up"});
  } else {
    job.fn.apply(this, job.args || []);
  }

  Array.prototype.push.apply(this.jobs, jobs);
};

Dissolve.prototype._exec_loop = function _exec_loop(job) {
  if (job.finished) {
    this.jobs.shift();
    return;
  }

  var jobs = this.jobs;
  this.jobs = [];

  if (job.name) {
    if (this.vars[job.name] === undefined) {
      this.vars[job.name] = [];
    }

    // private scope so _job doesn't get redefined later
    (function() {
      var _job = job;

      this.jobs.push({
        type: "tap",
        name: "__loop_temp",
        args: [job.finish],
        fn: job.fn,
      });

      this.jobs.push({
        type: "tap",
        fn: function() {
          if (!_job.cancelled) {
            this.vars[_job.name].push(this.vars.__loop_temp);
          }

          delete this.vars.__loop_temp;
        },
      });
    }).call(this);
  } else {
    this.jobs.push({
      type: "tap",
      args: [job.finish],
      fn: job.fn,
    });
  }
  Array.prototype.push.apply(this.jobs, jobs);
};

Dissolve.prototype._exec_string = function _exec_string(job, offset, length) {
  this.vars[job.name] = this._buffer.toString("utf8", offset, offset + length);
  this.jobs.shift();
};

Dissolve.prototype._exec_buffer = function _exec_buffer(job, offset, length) {
  this.vars[job.name] = this._buffer.slice(offset, offset+length);
  this.jobs.shift();
};

Dissolve.prototype._transform = function _transform(input, encoding, done) {
  var offset = 0;

  this._buffer.append(input);

  while (this.jobs.length) {
    var job = this.jobs[0];

    if (job.type === "down") {
      this._exec_down(job);
      continue;
    }

    if (job.type === "up") {
      this._exec_up(job);
      continue;
    }

    if (job.type === "tap") {
      this._exec_tap(job);
      continue;
    }

    if (job.type === "loop") {
      this._exec_loop(job);
      continue;
    }

    var length;
    if (typeof job.length === "string") {
      length = this.vars[job.length];
    } else {
      length = job.length;
    }

    if (this._buffer.length - offset < length) {
      break;
    }

    if (job.type === "buffer") {
      this._exec_buffer(job, offset, length);
      offset += length;
      continue;
    }

    if (job.type === "string") {
      this._exec_string(job, offset, length);
      offset += length;
      continue;
    }

    if (job.type === "skip") {
      this.jobs.shift();
      offset += length;
      continue;
    }

    switch (job.type) {
      case "int8le":  { this.vars[job.name] = this._buffer.readInt8(offset);  break; }
      case "uint8le": { this.vars[job.name] = this._buffer.readUInt8(offset); break; }
      case "int8be":  { this.vars[job.name] = this._buffer.readInt8(offset);  break; }
      case "uint8be": { this.vars[job.name] = this._buffer.readUInt8(offset); break; }
      case "int16le":  { this.vars[job.name] = this._buffer.readInt16LE(offset);  break; }
      case "uint16le": { this.vars[job.name] = this._buffer.readUInt16LE(offset); break; }
      case "int16be":  { this.vars[job.name] = this._buffer.readInt16BE(offset);  break; }
      case "uint16be": { this.vars[job.name] = this._buffer.readUInt16BE(offset); break; }
      case "int32le":  { this.vars[job.name] = this._buffer.readInt32LE(offset);  break; }
      case "uint32le": { this.vars[job.name] = this._buffer.readUInt32LE(offset); break; }
      case "int32be":  { this.vars[job.name] = this._buffer.readInt32BE(offset);  break; }
      case "uint32be": { this.vars[job.name] = this._buffer.readUInt32BE(offset); break; }
      case "int64le":  { this.vars[job.name] = (Math.pow(2, 32) * this._buffer.readInt32LE(offset + 4)) + ((this._buffer[offset + 4] & 0x80 === 0x80 ? 1 : -1) * this._buffer.readUInt32LE(offset)); break; }
      case "uint64le": { this.vars[job.name] = (Math.pow(2, 32) * this._buffer.readUInt32LE(offset + 4)) + this._buffer.readUInt32LE(offset); break; }
      case "int64be":  { this.vars[job.name] = (Math.pow(2, 32) * this._buffer.readInt32BE(offset)) + ((this._buffer[offset] & 0x80 === 0x80 ? 1 : -1) * this._buffer.readUInt32BE(offset + 4)); break; }
      case "uint64be": { this.vars[job.name] = (Math.pow(2, 32) * this._buffer.readUInt32BE(offset)) + this._buffer.readUInt32BE(offset + 4); break; }
      case "floatle":  { this.vars[job.name] = this._buffer.readFloatLE(offset);  break; }
      case "floatbe":  { this.vars[job.name] = this._buffer.readFloatBE(offset);  break; }
      case "doublele": { this.vars[job.name] = this._buffer.readDoubleLE(offset); break; }
      case "doublebe": { this.vars[job.name] = this._buffer.readDoubleBE(offset); break; }
      default: { return done(new Error("invalid job type")); }
    }

    this.jobs.shift();

    offset += length;
  }

  this._buffer.consume(offset);

  if (this.jobs.length === 0) {
    this.push(null);
  }

  return done();
};

[["float", 4], ["double", 8]].forEach(function(t) {
  ["", "le", "be"].forEach(function(e) {
    var id = [t[0], e].join(""),
        type = [t[0], e || "le"].join(""),
        length = t[1];

    Dissolve.prototype[id] = function(name) {
      this.jobs.push({
        type: type,
        length: length,
        name: name,
      });

      return this;
    };
  });
});

[8, 16, 32, 64].forEach(function(b) {
  ["", "u"].forEach(function(s) {
    ["", "le", "be"].forEach(function(e) {
      var id = [s, "int", b, e].join(""),
          type = [s, "int", b, e || "le"].join(""),
          length = b / 8;

      Dissolve.prototype[id] = function(name) {
        this.jobs.push({
          type: type,
          length: length,
          name: name,
        });

        return this;
      };
    });
  });
});

["tap", "loop"].forEach(function(e) {
  Dissolve.prototype[e] = function(name, fn) {
    if (typeof name === "function") {
      fn = name;
      name = null;
    }

    var job = {
      type: e,
      name: name,
      fn: fn,
    };

    if (e === "loop") {
      job.finish = function(cancel) {
        job.finished = true;
        job.cancelled = !!cancel;
      };
    }

    this.jobs.push(job);

    return this;
  };
});

["buffer", "string"].forEach(function(e) {
  Dissolve.prototype[e] = function(name, length) {
    this.jobs.push({
      type: e,
      name: name,
      length: length,
    });

    return this;
  };
});

["skip"].forEach(function(e) {
  Dissolve.prototype[e] = function(length) {
    this.jobs.push({
      type: e,
      length: length,
    });

    return this;
  };
});
